[ingester] Add metrics& healthcheck, rename Kafka cli flags

ledor473 opened the pull request:
## Which problem is this PR solving?
Resolves https://github.com/jaegertracing/jaeger/issues/1014
Resolves https://github.com/jaegertracing/jaeger/issues/1013
Resolves https://github.com/jaegertracing/jaeger/issues/1093

## Short description of the changes
- Add default healthcheck port (14270)
- Add a configurable metrics HTTP handler (default port is 14271)
- Make Kafka flags the same as in the Collector (prefix is `kafka` instead of `ingester`)

## Comments:
I wasn't sure if I should have refactor the following:
- Move the various Kafka flags under `/pkg`
 - I left it as is mainly because there's a mix of consumer vs producer flags and because the producer flags are under `plugins`

Also, should the `Parallelism` be under Kafka prefix? I felt it was more a `ingester` flags than a `kafka` one, but I don't have any strong opinion about it

yurishkuro commented on cmd/ingester/app/flags.go:
```diff
@@ -33,6 +33,8 @@ const (
 
 	// ConfigPrefix is a prefix fro the ingester flags
 	ConfigPrefix = "ingester"
+	// KafkaConfigPrefix is a prefix fro the ingester flags
```
s/fro/for/

yurishkuro commented on cmd/ingester/app/flags.go:
```diff
@@ -33,6 +33,8 @@ const (
 
 	// ConfigPrefix is a prefix fro the ingester flags
```
s/fro/for/

yurishkuro commented on cmd/ingester/app/flags.go:
```diff
@@ -54,27 +58,33 @@ const (
 	DefaultParallelism = 1000
 	// DefaultEncoding is the default span encoding
 	DefaultEncoding = EncodingProto
+	// DefaultHTTPPort is the default HTTP port
```
add: "e.g. for /metrics" (I was at first confused why we need http port)

ledor473 commented:
I'm not sure what would be the best way to fix that problem in the CI. The flags is indeed redefined, but they usually run in different Go applications
```
--- FAIL: TestKafkaStorage (0.00s)
panic: flag redefined: kafka.brokers [recovered]
	panic: flag redefined: kafka.brokers
```

I've also fixed the Ingester flag tests and moved `ingester.encoding` to `kafka.encoding` because it's like that in Collector

yurishkuro commented:
> The flags is indeed redefined, but they usually run in different Go applications

This sounds like an issue with the tests. We should be able to create a new instance of the FlagSet in the unit test instead of depending on the global one.

ledor473 commented on plugin/storage/integration/kafka_test.go:
```diff
@@ -48,7 +48,7 @@ func (s *KafkaIntegrationTestSuite) initialize() error {
 
 	f := kafka.NewFactory()
-	v, command := config.Viperize(f.AddFlags, app.AddFlags)
+	v, command := config.Viperize(app.AddFlags)
```
@yurishkuro the change here fixed the test. `app.AddFlags` already contained all flags needed by `f.InitFromViper(v)`

yurishkuro replied:
is kafka factory's AddFlags() no-op then?looks a bit odd

jpkrohling commented:
@ledor473 do you still have anything to work here, or can I merge it ?

ledor473 commented:
@jpkrohling It's ready to merge! I'll rebase it real quick

jpkrohling commented:
Once `codecov` turns green, this will be merged.

yurishkuro commented:
I restarted test coverage job

objectiser commented on cmd/ingester/app/flags.go:
```diff
@@ -54,44 +58,56 @@ const (
+	// DefaultHTTPPort is the default HTTP port (e.g. for /metrics)
+	DefaultHTTPPort = 14271
+	// IngesterDefaultHealthCheckHTTPPort is the default HTTP Port for health check
+	IngesterDefaultHealthCheckHTTPPort = 14270
```
nit: Does this need to have the `Ingester` prefix - as the other constants don't?

ledor473 replied:
I've taken the variable names from here and both are prefixed: 

https://github.com/jaegertracing/jaeger/blob/bbf8b7a1eedefa79eefeaef894bbf10928d0a125/cmd/query/app/flags.go#L29
https://github.com/jaegertracing/jaeger/blob/e93c37503532ab7a18a3bdf21eee98194616017f/cmd/collector/app/builder/builder_flags.go#L33

That being said, there's a prefix on the other variable as well

objectiser commented on cmd/ingester/app/flags.go:
```diff
@@ -31,18 +31,22 @@ const (
-	// ConfigPrefix is a prefix fro the ingester flags
+	// ConfigPrefix is a prefix for the ingester flags
 	ConfigPrefix = "ingester"
+	// KafkaConfigPrefix is a prefix for the ingester flags
```
"for the Kafka flags"

objectiser commented on cmd/ingester/app/flags.go:
```diff
@@ -54,44 +58,56 @@ const (
-		ConfigPrefix+SuffixEncoding,
-		DefaultEncoding,
-		fmt.Sprintf(`The encoding of spans ("%s" or "%s") consumed from kafka`, EncodingProto, EncodingJSON))
+	flagSet.Int(
```
Should there also be a flag to allow the health check port to be specified?

pavolloffay replied:
The flag is defined in healthcheck package

ledor473 replied:
As @pavolloffay said, it's set using flags.SetDefaultHealthCheckPort

pavolloffay commented:
@ledor473 could you please fix DCO of the last commit? I will merge afterwards.

pavolloffay commented:
thanks @ledor473

ledor473 commented:
@yurishkuro @pavolloffay @objectiser I've just realized that if you run the Ingester with `SPAN_STORAGE_TYPE=elasticsearch,kafka`, the Ingester crash with the following:
```
WARNING: multiple span storage types have been specified. Only the first type (elasticsearch) will be used for reading and archiving.

flag redefined: kafka.brokers
panic: flag redefined: kafka.brokers

goroutine 1 [running]:
flag.(*FlagSet).Var(0xc000080d80, 0xefd9a0, 0xc00018d080, 0xe04f6e, 0xd, 0xe2cb58, 0x4b)
	/home/travis/.gimme/versions/go1.11.1.linux.amd64/src/flag/flag.go:805 +0x529
flag.(*FlagSet).StringVar(0xc000080d80, 0xc00018d080, 0xe04f6e, 0xd, 0xe05426, 0xe, 0xe2cb58, 0x4b)
	/home/travis/.gimme/versions/go1.11.1.linux.amd64/src/flag/flag.go:708 +0x8a
flag.(*FlagSet).String(0xc000080d80, 0xe04f6e, 0xd, 0xe05426, 0xe, 0xe2cb58, 0x4b, 0xe00e66)
	/home/travis/.gimme/versions/go1.11.1.linux.amd64/src/flag/flag.go:721 +0x8b
github.com/jaegertracing/jaeger/cmd/ingester/app.AddFlags(0xc000080d80)
	/home/travis/gopath/src/github.com/jaegertracing/jaeger/cmd/ingester/app/flags.go:84 +0x6e
github.com/jaegertracing/jaeger/pkg/config.AddFlags(0xc00015a0f0, 0xc000186240, 0xc000159f28, 0x5, 0x5, 0xc00003a042, 0xd)
	/home/travis/gopath/src/github.com/jaegertracing/jaeger/pkg/config/config.go:36 +0x56
main.main()
	/home/travis/gopath/src/github.com/jaegertracing/jaeger/cmd/ingester/main.go:138 +0x59f
```
